---
title: "TCGA_ML"
author: "Maria Dermit"
date: "06/07/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
rm(list = ls())
```

# Importing Library 
```{r message=FALSE, warning=FALSE}
library("TCGAbiolinks")
library("glmnet")
library("factoextra")
library("FactoMineR")
library("caret")
library("SummarizedExperiment")
library("ggplot2")
library("RColorBrewer")
library("gProfileR")
library("genefilter")
library("GenomicDataCommons")
library("GenomeInfoDbData")
library("keras")
library("tensorflow")
library("dplyr")

```

# Import data from TCGA 
```{r}
GDCprojects <- TCGAbiolinks::getGDCprojects()
#Let's see what data is available for prostate adenocarcinoma
TCGAbiolinks:::getProjectSummary("TCGA-PRAD")
```

# Configuring Dataset from TCGA 

```{r}
query_TCGA <- GDCquery(
  project = "TCGA-PRAD",
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  experimental.strategy = "RNA-Seq",
  barcode = c("TCGA-*")) # A list of barcodes to filter the files to download

class(query_TCGA)

GDCdownload(query = query_TCGA, method = "api", files.per.chunk = 100)

query_TCGA <- query_TCGA[,colSums(is.na(query_TCGA))<nrow(query_TCGA)]
tcga_data <- GDCprepare(query_TCGA,summarizedExperiment = TRUE)

class(tcga_data)

saveRDS(tcga_data,file = "TCGA.rds")
tcga_data <- readRDS("TCGA.rds")
```

# Data structring and Preparations
```{r}
sedf <- tcga_data

table(sedf@colData@listData$sample_type)

#there is only one metastatic sample, so let's remove itt

sedf <- sedf[,sedf@colData@listData$sample_type != "Metastatic"]

geneslist <- sedf@rowRanges$gene_id
samplelist <- sedf@colData@listData$sample

expr <- sedf@assays@data@listData$tpm_unstrand
dim(expr)

rownames(expr) <- geneslist
colnames(expr) <- sedf@colData@listData$sample

#Remove house-keeping genes with more TPM 2000
drop <- apply(expr,1,min) > 2000
expr_filtered <- expr[!drop,]

#Remove low-expressing genes
drop2 <- apply(expr_filtered ,1,max) < 100
expr_filtered <- expr_filtered[!drop2,]

dim(expr_filtered)

rm(expr)
#rm(tcga_data)
gc()
```

# Visualized Data relationship

## limma plotMDS()
```{r}
library(limma)
library(edgeR)

dge <- DGEList(t(expr_filtered))
#plot multidimensional data
plotMDS.default(dge,cex = 0.4,col=tcga_data@colData@listData$sample_type)
```


# Min-Max (0-1) + Log2 Normalization of the training data

```{r}
#normalize <- function(x)
#    {
#    return((x- min(x)) /(max(x)-min(x)))
#}

#copydata <- function(x){
#  y <- rbind(x,x,x,x,x)
#  return(y)}

#need to transpose so gene becomes the column and sample becomes the row
train_data <- unlist(t(expr_filtered))
train_data <- log2(train_data+0.5)
#train_data <- train_data %>% normalize() %>% copydata()
dim(train_data) <- dim(t(expr_filtered))

dim(train_data)

#lets plot histograms for the first three samples 
hist(train_data[1,])
hist(train_data[2,])
hist(train_data[3,])

#they seem to be normally distributed

```
## Princple Component Analysis
```{r}
pca.gse <- PCA(train_data,graph = FALSE)
table(sedf@colData@listData[["secondary_gleason_grade"]])

#Visualize Principal Component Analysis
fviz_pca_ind(pca.gse, geom = "point", col.ind = sedf@colData@listData[["secondary_gleason_grade"]])

#in the PCA space these three patterns are hard to separate. 
```

# Preparing Training Label + Converting Catergorical data to One-Hot Encoding 

```{r}
train_label <- sedf@colData@listData[["secondary_gleason_grade"]]
#one hot encoding
train_label <- train_label %>% as.factor() %>% as.numeric()
train_label <- train_label - 1
#restructure into 2 dim matrix
dim(train_label) <- c(dim(expr_filtered)[2], 1)

train_label <- keras::to_categorical(train_label, num_classes = 3)

pheatmap(train_label,cluster_cols = FALSE,cluster_rows = FALSE)

#we already knew that Pattern 4 is the most common secondary_gleason_grade
```


# Construcing Neural Network 

```{r}
rm(model)
#how many genes there are
dim(train_data)[2]
```


```{r}
# netxwork has 5 dense layers. First layer has  4096 neurons 
NNarray <- c(4096,2048, 1024, 512) 

model <- keras_model_sequential() %>% #sequential layer 
  layer_dense( #first dense later
    units = NNarray[1],
    activation = "relu",
    input_shape = dim(train_data)[2]
  ) %>%
  layer_dense(units = NNarray[2], activation = "relu") %>%
  layer_dense(units = NNarray[3], activation = "relu") %>%
  layer_dense(units = NNarray[4], activation = "relu") %>%
  layer_dense(units = dim(train_label)[2], activation = "sigmoid") #last layer is the output. Classification.

summary(model)

#compile the model
model %>% compile(optimizer = 'sgd', loss = "binary_crossentropy", metrics = c('accuracy'))

#train the model
history <- model %>%
  fit(
    x = train_data,
    y = train_label,
    epochs = 400, #200*552  is how many times the network sees the data 
    use_multiprocessing = TRUE,
    batch_size = dim(train_data)[1]/25,
    #validation_split = 0.1
  )
```


```{r}
save_model_hdf5(model,'model1.hdf5')

model <- keras::load_model_hdf5('model1.hdf5')

evaluate(model, train_data, train_label)
#0.7355072 of the time the network is able to choose the correct pattern based on the input 

```

# Extraction of Weights and Bias 
```{r}
#we are only extracting the weights and bias from layer 1. There may be noise coming from this layer that could be filtered off from layer 2,3,4..
weight <- as.matrix(model$layers[[1]]$weights[[1]])
dim(weight)
bias   <- as.matrix(model$layers[[1]]$weights[[2]])
dim(bias) #there are 4096 nodes

expr_filtered <- as.matrix(expr_filtered)

dim(weight)
dim(expr_filtered)

rownames(weight) <- rownames(expr_filtered)

hist(weight[56,])
hist(bias)

#pheatmap(weight,cluster_rows = FALSE,cluster_cols = FALSE)
```
# Isolation of Gene of Interests (GOI)
```{r}
#which genes are causing neurons in layer one to fire?

GOI <- c()

input_data  <- train_data

#lets use the first 100 patients only. Assumption is that genes are not so different across patients
samplerate <- sample(1:nrow(input_data))[1:100]

for(j in samplerate) {
#for(j in 1:5) {
  sample1 <- train_data[j, ]
  
  total_weights <- weight * sample1
  total_weights_bias <- colSums(total_weights) + bias
  total_weights_bias <- as.data.frame(total_weights_bias)
  total_weights_bias <- cbind(seq(1, NNarray[1]), total_weights_bias)
  
  #top nodes based on Relu more than 1
  top_nodes <- as.data.frame(total_weights_bias[total_weights_bias[,2] > 1, ])
  #genes that contribute the most to those nodes
  goodnodes <- top_nodes$V1
  genes_to_goodnodes <- as.data.frame(total_weights[, goodnodes])
  #genes_to_goodnodes has all the genes but only a selection number of nodes
  
  rm(sample1)
  rm(total_weights)
  rm(total_weights_bias)
  rm(top_nodes)
  
  if(ncol(genes_to_goodnodes)>0){
    #go to nodes one by one and extract genes
    for(i in 1:ncol(genes_to_goodnodes)){
      temp <- genes_to_goodnodes[,i]
      names(temp) <- rownames(genes_to_goodnodes)
      temp <- sort(temp,decreasing = TRUE)
      temp <- temp[1:10]
      temp <- names(temp)
      
      GOI <- c(GOI, temp)
      rm(temp)
      }
    }
  
  print(paste("j=",j))
}

  sample1 <- train_data[j, ]
  
  total_weights <- weight * sample1
  total_weights_bias <- colSums(total_weights) + bias
  total_weights_bias <- as.data.frame(total_weights_bias)
  total_weights_bias <- cbind(seq(1, NNarray[1]), total_weights_bias)
  

GOI_table <- as.data.frame(table(GOI))
GOI_table <- GOI_table %>% arrange(desc(Freq))
GOI_table <- GOI_table[GOI_table$Freq>1,]
#genes that are causing node to fire 
head(GOI_table)
GOI_list <- unique(GOI)

dir.create("Output", showWarnings = FALSE)
write.csv(GOI_table, file = "Output/GOI_LIST.csv")
```

# Converting EnsembleID to EntrezID (GOI)
```{r}
gc()
genelist <- sub("[.][0-9]*", "", GOI_list)

library(clusterProfiler)
library(org.Hs.eg.db)

#clusterProfiler needs ENTREZID, so we need this conversion
new_genelist <- clusterProfiler::bitr(
  genelist,
  fromType = "ENSEMBL",
  toType = "ENTREZID",
  OrgDb = org.Hs.eg.db,
  drop = TRUE
)

nrow(new_genelist) #because some failed to transform
head(new_genelist)

```

# Converting EnsembleID to EntrezID (All Genes)
```{r}
fullgenelist <- rownames(expr_filtered)
fullgenelist <- sub("[.][0-9]*", "", fullgenelist)

new_fullgenelist <- bitr(
  fullgenelist,
  fromType = "ENSEMBL",
  toType = "ENTREZID",
  OrgDb = org.Hs.eg.db,
  drop = TRUE
)

head(new_fullgenelist)
nrow(new_fullgenelist)

```
# Gene Set Enrichment using enrichGO() from Cluster profiler - Biological Process 
```{r}
gene <- new_genelist$ENTREZID
fullgene <- new_fullgenelist$ENTREZID

DOSEgenelist <- data(geneList, package = "DOSE")

library(org.Hs.eg.db)

ggo_BP <- groupGO(
  gene     = gene,
  OrgDb    = org.Hs.eg.db,
  ont      = "BP",
  level    = 3,
  readable = TRUE
)

ego_BP <- enrichGO(
  gene       = gene,
  universe      = fullgene,
  OrgDb         = org.Hs.eg.db,
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

head(ego_BP, 10)
goplot(ego_BP)

ego_BP_df <- ego_BP@result
write.csv(ego_BP_df, file= "Output/ego_BP_df.csv")
```



# Gene Set Enrichment using enrichGO() from Cluster profiler - Molecular Processes 
```{r}
ggo_MF <- groupGO(
  gene     = gene,
  OrgDb    = org.Hs.eg.db,
  ont      = "MF",
  level    = 3,
  readable = TRUE
)


ego_MF <- enrichGO(
  gene       = gene,
  universe      = fullgene,
  OrgDb         = org.Hs.eg.db,
  ont           = "MF",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

head(ego_MF, 20)
goplot(ego_MF)

ego_MF_df <- ego_MF@result
write.csv(ego_MF_df, file= "Output/ego_MF_df.csv")

```
# Gene Set Enrichment using enrichGO() from Cluster profiler - Cellular Categories  
```{r}
ggo_MF <- groupGO(
  gene     = gene,
  OrgDb    = org.Hs.eg.db,
  ont      = "CC",
  level    = 3,
  readable = TRUE
)


ego_CC <- enrichGO(
  gene       = gene,
  universe      = fullgene,
  OrgDb         = org.Hs.eg.db,
  ont           = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.01,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

head(ego_CC, 20)
goplot(ego_CC)

ego_cc_df <- ego_CC@result
write.csv(ego_cc_df, file= "Output/ego_cc_df.csv")
```
